File: Clients/DotNet/ConduitLLM.CoreClient/src/Models/Metrics.cs
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

namespace ConduitLLM.CoreClient.Models;

/// <summary>
/// Comprehensive snapshot of system metrics at a point in time.
/// </summary>
public class MetricsSnapshot
{
    /// <summary>
    /// Gets or sets the timestamp when these metrics were captured.
    /// </summary>
    public DateTime Timestamp { get; set; } = DateTime.UtcNow;

    /// <summary>
    /// Gets or sets HTTP request and response metrics.
    /// </summary>
    public HttpMetrics Http { get; set; } = new();

    /// <summary>
    /// Gets or sets infrastructure component metrics.
    /// </summary>
    public InfrastructureMetrics Infrastructure { get; set; } = new();

    /// <summary>
    /// Gets or sets business logic metrics.
    /// </summary>
    public BusinessMetrics Business { get; set; } = new();

    /// <summary>
    /// Gets or sets provider health status information.
    /// </summary>
    public List<ProviderHealthStatus> ProviderHealth { get; set; } = new();

    /// <summary>
    /// Gets or sets system resource metrics.
    /// </summary>
    public SystemMetrics System { get; set; } = new();
}

/// <summary>
/// HTTP request and response metrics.
/// </summary>
public class HttpMetrics
{
    /// <summary>
    /// Gets or sets the number of requests per second.
    /// </summary>
    public double RequestsPerSecond { get; set; }

    /// <summary>
    /// Gets or sets the current number of active requests.
    /// </summary>
    public int ActiveRequests { get; set; }

    /// <summary>
    /// Gets or sets request rates per endpoint.
    /// </summary>
    public Dictionary<string, double> EndpointRequestRates { get; set; } = new();

    /// <summary>
    /// Gets or sets response time percentiles.
    /// </summary>
    public ResponseTimeMetrics ResponseTimes { get; set; } = new();

    /// <summary>
    /// Gets or sets the overall error rate as a percentage.
    /// </summary>
    public double ErrorRate { get; set; }

    /// <summary>
    /// Gets or sets counts of responses by HTTP status code.
    /// </summary>
    public Dictionary<string, int> StatusCodeCounts { get; set; } = new();

    /// <summary>
    /// Gets or sets the number of rate limit hits per minute.
    /// </summary>
    public double RateLimitHitsPerMinute { get; set; }
}

/// <summary>
/// Response time metrics with percentiles.
/// </summary>
public class ResponseTimeMetrics
{
    /// <summary>
    /// Gets or sets the 50th percentile (median) response time in milliseconds.
    /// </summary>
    public double P50 { get; set; }

    /// <summary>
    /// Gets or sets the 90th percentile response time in milliseconds.
    /// </summary>
    public double P90 { get; set; }

    /// <summary>
    /// Gets or sets the 95th percentile response time in milliseconds.
    /// </summary>
    public double P95 { get; set; }

    /// <summary>
    /// Gets or sets the 99th percentile response time in milliseconds.
    /// </summary>
    public double P99 { get; set; }

    /// <summary>
    /// Gets or sets the average response time in milliseconds.
    /// </summary>
    public double Average { get; set; }
}

/// <summary>
/// Infrastructure component metrics.
/// </summary>
public class InfrastructureMetrics
{
    /// <summary>
    /// Gets or sets database metrics.
    /// </summary>
    public DatabaseMetrics Database { get; set; } = new();

    /// <summary>
    /// Gets or sets Redis cache metrics.
    /// </summary>
    public RedisMetrics Redis { get; set; } = new();

    /// <summary>
    /// Gets or sets RabbitMQ message queue metrics.
    /// </summary>
    public RabbitMQMetrics? RabbitMQ { get; set; }

    /// <summary>
    /// Gets or sets SignalR real-time communication metrics.
    /// </summary>
    public SignalRMetrics SignalR { get; set; } = new();
}

/// <summary>
/// Database connection and performance metrics.
/// </summary>
public class DatabaseMetrics
{
    /// <summary>
    /// Gets or sets the number of active database connections.
    /// </summary>
    public int ActiveConnections { get; set; }

    /// <summary>
    /// Gets or sets the maximum allowed database connections.
    /// </summary>
    public int MaxConnections { get; set; }

    /// <summary>
    /// Gets or sets the connection pool utilization percentage.
    /// </summary>
    public double PoolUtilization => MaxConnections > 0 ? (double)ActiveConnections / MaxConnections * 100 : 0;

    /// <summary>
    /// Gets or sets the average query duration in milliseconds.
    /// </summary>
    public double AverageQueryDuration { get; set; }

    /// <summary>
    /// Gets or sets the number of queries per second.
    /// </summary>
    public double QueriesPerSecond { get; set; }

    /// <summary>
    /// Gets or sets the number of failed queries per minute.
    /// </summary>
    public double FailedQueriesPerMinute { get; set; }
}

/// <summary>
/// Redis cache metrics.
/// </summary>
public class RedisMetrics
{
    /// <summary>
    /// Gets or sets the Redis memory usage in megabytes.
    /// </summary>
    public double MemoryUsageMB { get; set; }

    /// <summary>
    /// Gets or sets the total number of keys in Redis.
    /// </summary>
    public long KeyCount { get; set; }

    /// <summary>
    /// Gets or sets the cache hit rate as a percentage.
    /// </summary>
    public double HitRate { get; set; }

    /// <summary>
    /// Gets or sets the average Redis operation latency in milliseconds.
    /// </summary>
    public double AverageLatency { get; set; }

    /// <summary>
    /// Gets or sets the number of operations per second.
    /// </summary>
    public double OperationsPerSecond { get; set; }
}

/// <summary>
/// RabbitMQ message queue metrics.
/// </summary>
public class RabbitMQMetrics
{
    /// <summary>
    /// Gets or sets queue depths by queue name.
    /// </summary>
    public Dictionary<string, int> QueueDepths { get; set; } = new();

    /// <summary>
    /// Gets or sets message rates by queue name.
    /// </summary>
    public Dictionary<string, double> MessageRates { get; set; } = new();

    /// <summary>
    /// Gets or sets consumer counts by queue name.
    /// </summary>
    public Dictionary<string, int> ConsumerCounts { get; set; } = new();

    /// <summary>
    /// Gets or sets the total number of unacknowledged messages.
    /// </summary>
    public int UnacknowledgedMessages { get; set; }

    /// <summary>
    /// Gets or sets the connection count.
    /// </summary>
    public int ConnectionCount { get; set; }
}

/// <summary>
/// SignalR real-time communication metrics.
/// </summary>
public class SignalRMetrics
{
    /// <summary>
    /// Gets or sets the number of active SignalR connections.
    /// </summary>
    public int ActiveConnections { get; set; }

    /// <summary>
    /// Gets or sets the message rate per second.
    /// </summary>
    public double MessagesPerSecond { get; set; }

    /// <summary>
    /// Gets or sets the average message processing time in milliseconds.
    /// </summary>
    public double AverageProcessingTime { get; set; }

    /// <summary>
    /// Gets or sets connection counts by hub type.
    /// </summary>
    public Dictionary<string, int> ConnectionsByHub { get; set; } = new();
}

/// <summary>
/// Business logic and application metrics.
/// </summary>
public class BusinessMetrics
{
    /// <summary>
    /// Gets or sets the number of active virtual keys.
    /// </summary>
    public int ActiveVirtualKeys { get; set; }

    /// <summary>
    /// Gets or sets the total number of requests per minute.
    /// </summary>
    public double TotalRequestsPerMinute { get; set; }

    /// <summary>
    /// Gets or sets cost-related metrics.
    /// </summary>
    public CostMetrics Cost { get; set; } = new();

    /// <summary>
    /// Gets or sets model usage statistics.
    /// </summary>
    public List<ModelUsageStats> ModelUsage { get; set; } = new();

    /// <summary>
    /// Gets or sets virtual key statistics.
    /// </summary>
    public List<VirtualKeyStats> VirtualKeyStats { get; set; } = new();
}

/// <summary>
/// Cost-related metrics.
/// </summary>
public class CostMetrics
{
    /// <summary>
    /// Gets or sets the cost per minute in USD.
    /// </summary>
    public decimal CostPerMinute { get; set; }

    /// <summary>
    /// Gets or sets cost breakdown by provider.
    /// </summary>
    public Dictionary<string, decimal> CostByProvider { get; set; } = new();

    /// <summary>
    /// Gets or sets the average cost per request in USD.
    /// </summary>
    public decimal AverageCostPerRequest { get; set; }

    /// <summary>
    /// Gets or sets the total spend in the current billing period.
    /// </summary>
    public decimal TotalSpendCurrentPeriod { get; set; }
}

/// <summary>
/// Model usage statistics.
/// </summary>
public class ModelUsageStats
{
    /// <summary>
    /// Gets or sets the model name.
    /// </summary>
    [Required]
    public string ModelName { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the provider name.
    /// </summary>
    [Required]
    public string ProviderName { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the number of requests per minute.
    /// </summary>
    public double RequestsPerMinute { get; set; }

    /// <summary>
    /// Gets or sets the total tokens processed per minute.
    /// </summary>
    public long TokensPerMinute { get; set; }

    /// <summary>
    /// Gets or sets the average response time in milliseconds.
    /// </summary>
    public double AverageResponseTime { get; set; }

    /// <summary>
    /// Gets or sets the error rate as a percentage.
    /// </summary>
    public double ErrorRate { get; set; }

    /// <summary>
    /// Gets or sets the cost per minute for this model.
    /// </summary>
    public decimal CostPerMinute { get; set; }
}

/// <summary>
/// Virtual key performance statistics.
/// </summary>
public class VirtualKeyStats
{
    /// <summary>
    /// Gets or sets the virtual key identifier.
    /// </summary>
    [Required]
    public string VirtualKeyId { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the virtual key name.
    /// </summary>
    public string? VirtualKeyName { get; set; }

    /// <summary>
    /// Gets or sets the current spend amount.
    /// </summary>
    public decimal CurrentSpend { get; set; }

    /// <summary>
    /// Gets or sets the budget utilization percentage.
    /// </summary>
    public double BudgetUtilization { get; set; }

    /// <summary>
    /// Gets or sets the request rate per minute.
    /// </summary>
    public double RequestsPerMinute { get; set; }

    /// <summary>
    /// Gets or sets the average cost per request.
    /// </summary>
    public decimal AverageCostPerRequest { get; set; }

    /// <summary>
    /// Gets or sets the error rate as a percentage.
    /// </summary>
    public double ErrorRate { get; set; }
}

/// <summary>
/// System resource metrics.
/// </summary>
public class SystemMetrics
{
    /// <summary>
    /// Gets or sets the CPU usage percentage.
    /// </summary>
    public double CpuUsagePercent { get; set; }

    /// <summary>
    /// Gets or sets the memory usage in megabytes.
    /// </summary>
    public double MemoryUsageMB { get; set; }

    /// <summary>
    /// Gets or sets the current thread count.
    /// </summary>
    public int ThreadCount { get; set; }

    /// <summary>
    /// Gets or sets the garbage collector memory usage in megabytes.
    /// </summary>
    public double GcMemoryMB { get; set; }

    /// <summary>
    /// Gets or sets the system uptime.
    /// </summary>
    public TimeSpan Uptime { get; set; }

    /// <summary>
    /// Gets or sets the process start time.
    /// </summary>
    public DateTime ProcessStartTime { get; set; }
}

/// <summary>
/// Provider health status information.
/// </summary>
public class ProviderHealthStatus
{
    /// <summary>
    /// Gets or sets the provider name.
    /// </summary>
    [Required]
    public string ProviderName { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the current health status.
    /// </summary>
    [Required]
    public string Status { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the timestamp of the last successful request.
    /// </summary>
    public DateTime? LastSuccessfulRequest { get; set; }

    /// <summary>
    /// Gets or sets the error rate as a percentage.
    /// </summary>
    public double ErrorRate { get; set; }

    /// <summary>
    /// Gets or sets the average latency in milliseconds.
    /// </summary>
    public double AverageLatency { get; set; }

    /// <summary>
    /// Gets or sets the number of available models for this provider.
    /// </summary>
    public int AvailableModels { get; set; }

    /// <summary>
    /// Gets or sets whether the provider is enabled.
    /// </summary>
    public bool IsEnabled { get; set; }

    /// <summary>
    /// Gets whether the provider is currently healthy.
    /// </summary>
    [JsonIgnore]
    public bool IsHealthy => Status.Equals("healthy", StringComparison.OrdinalIgnoreCase);
}

/// <summary>
/// Request for historical metrics data.
/// </summary>
public class HistoricalMetricsRequest
{
    /// <summary>
    /// Gets or sets the start time for the metrics query.
    /// </summary>
    [Required]
    public DateTime StartTime { get; set; }

    /// <summary>
    /// Gets or sets the end time for the metrics query.
    /// </summary>
    [Required]
    public DateTime EndTime { get; set; }

    /// <summary>
    /// Gets or sets the list of metric names to retrieve.
    /// </summary>
    public List<string> MetricNames { get; set; } = new();

    /// <summary>
    /// Gets or sets the aggregation interval (e.g., "1m", "5m", "1h").
    /// </summary>
    public string Interval { get; set; } = "1m";

    /// <summary>
    /// Gets or sets additional filters for the query.
    /// </summary>
    public Dictionary<string, string>? Filters { get; set; }
}

/// <summary>
/// Response containing historical metrics data.
/// </summary>
public class HistoricalMetricsResponse
{
    /// <summary>
    /// Gets or sets the start time of the data range.
    /// </summary>
    public DateTime StartTime { get; set; }

    /// <summary>
    /// Gets or sets the end time of the data range.
    /// </summary>
    public DateTime EndTime { get; set; }

    /// <summary>
    /// Gets or sets the aggregation interval used.
    /// </summary>
    public string Interval { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the time series data.
    /// </summary>
    public List<MetricsSeries> Series { get; set; } = new();
}

/// <summary>
/// Time series data for a specific metric.
/// </summary>
public class MetricsSeries
{
    /// <summary>
    /// Gets or sets the metric name.
    /// </summary>
    [Required]
    public string MetricName { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the time series data points.
    /// </summary>
    public List<MetricsDataPoint> DataPoints { get; set; } = new();

    /// <summary>
    /// Gets or sets metadata about the metric.
    /// </summary>
    public Dictionary<string, string> Metadata { get; set; } = new();
}

/// <summary>
/// A single data point in a time series.
/// </summary>
public class MetricsDataPoint
{
    /// <summary>
    /// Gets or sets the timestamp of this data point.
    /// </summary>
    public DateTime Timestamp { get; set; }

    /// <summary>
    /// Gets or sets the metric value.
    /// </summary>
    public double Value { get; set; }

    /// <summary>
    /// Gets or sets additional labels for this data point.
    /// </summary>
    public Dictionary<string, string>? Labels { get; set; }
}

/// <summary>
/// Metric alert information.
/// </summary>
public class MetricAlert
{
    /// <summary>
    /// Gets or sets the unique alert identifier.
    /// </summary>
    [Required]
    public string Id { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the alert severity level.
    /// </summary>
    [Required]
    public string Severity { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the name of the metric that triggered the alert.
    /// </summary>
    [Required]
    public string MetricName { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the alert message.
    /// </summary>
    [Required]
    public string Message { get; set; } = string.Empty;

    /// <summary>
    /// Gets or sets the current value of the metric.
    /// </summary>
    public double CurrentValue { get; set; }

    /// <summary>
    /// Gets or sets the threshold that was exceeded.
    /// </summary>
    public double Threshold { get; set; }

    /// <summary>
    /// Gets or sets when the alert was triggered.
    /// </summary>
    public DateTime TriggeredAt { get; set; }

    /// <summary>
    /// Gets or sets whether the alert is currently active.
    /// </summary>
    public bool IsActive { get; set; }

    /// <summary>
    /// Gets or sets additional context about the alert.
    /// </summary>
    public Dictionary<string, string>? Context { get; set; }
}

File: SDKs/Node/Core/src/models/metrics.ts
import { ProviderType } from './providerType';

/**
 * Comprehensive metrics snapshot from the API
 */
export interface MetricsSnapshot {
  /** Timestamp when metrics were collected */
  timestamp: Date;
  /** HTTP performance metrics */
  http: HttpMetrics;
  /** Business metrics including costs and usage */
  business: BusinessMetrics;
  /** System resource metrics */
  system: SystemMetrics;
  /** Infrastructure component metrics */
  infrastructure: InfrastructureMetrics;
  /** Provider health status for all providers */
  providerHealth: ProviderHealthStatus[];
}

/**
 * HTTP performance metrics
 */
export interface HttpMetrics {
  /** Current requests per second */
  requestsPerSecond: number;
  /** Number of active requests */
  activeRequests: number;
  /** Error rate percentage */
  errorRate: number;
  /** Response time statistics */
  responseTimes: ResponseTimeMetrics;
}

/**
 * Response time metrics
 */
export interface ResponseTimeMetrics {
  /** Average response time in milliseconds */
  average: number;
  /** 50th percentile response time */
  p50: number;
  /** 95th percentile response time */
  p95: number;
  /** 99th percentile response time */
  p99: number;
  /** Minimum response time */
  min: number;
  /** Maximum response time */
  max: number;
}

/**
 * Business metrics including costs and usage
 */
export interface BusinessMetrics {
  /** Number of active virtual keys */
  activeVirtualKeys: number;
  /** Total requests per minute across all keys */
  totalRequestsPerMinute: number;
  /** Cost-related metrics */
  cost: CostMetrics;
  /** Model usage statistics */
  modelUsage: ModelUsageStats[];
  /** Virtual key statistics */
  virtualKeyStats: VirtualKeyStats[];
}

/**
 * Cost-related metrics
 */
export interface CostMetrics {
  /** Cost per minute */
  costPerMinute: number;
  /** Average cost per request */
  averageCostPerRequest: number;
  /** Total cost for current period */
  totalCost: number;
}

/**
 * Model usage statistics
 */
export interface ModelUsageStats {
  /** Model name */
  modelName: string;
  /** Requests per minute for this model */
  requestsPerMinute: number;
  /** Total requests for this model */
  totalRequests: number;
  /** Average cost per request */
  averageCost: number;
  /** Error rate for this model */
  errorRate: number;
}

/**
 * Virtual key statistics
 */
export interface VirtualKeyStats {
  /** Virtual key ID */
  keyId: string;
  /** Key name */
  keyName: string;
  /** Current spend amount */
  currentSpend: number;
  /** Budget limit */
  budgetLimit?: number;
  /** Requests per minute */
  requestsPerMinute: number;
  /** Whether the key is enabled */
  isEnabled: boolean;
}

/**
 * System resource metrics
 */
export interface SystemMetrics {
  /** CPU usage percentage */
  cpuUsagePercent: number;
  /** Memory usage in MB */
  memoryUsageMB: number;
  /** Available memory in MB */
  availableMemoryMB: number;
  /** System uptime */
  uptime: string;
  /** Number of CPU cores */
  cpuCores: number;
  /** Thread count */
  threadCount: number;
}

/**
 * Infrastructure component metrics
 */
export interface InfrastructureMetrics {
  /** Database metrics */
  database: DatabaseMetrics;
  /** Redis cache metrics */
  redis: RedisMetrics;
  /** SignalR metrics */
  signalR: SignalRMetrics;
  /** RabbitMQ metrics (if configured) */
  rabbitmq?: RabbitMQMetrics;
}

/**
 * Database connection pool metrics
 */
export interface DatabaseMetrics {
  /** Number of active connections */
  activeConnections: number;
  /** Maximum number of connections */
  maxConnections: number;
  /** Pool utilization percentage */
  poolUtilization: number;
  /** Average connection acquisition time */
  averageConnectionAcquisitionTime: number;
  /** Number of failed connections */
  failedConnections: number;
}

/**
 * Redis cache metrics
 */
export interface RedisMetrics {
  /** Cache hit rate percentage */
  hitRate: number;
  /** Number of cache hits */
  hits: number;
  /** Number of cache misses */
  misses: number;
  /** Memory usage in MB */
  memoryUsageMB: number;
  /** Number of connected clients */
  connectedClients: number;
  /** Operations per second */
  operationsPerSecond: number;
}

/**
 * SignalR metrics
 */
export interface SignalRMetrics {
  /** Number of active connections */
  activeConnections: number;
  /** Messages per second */
  messagesPerSecond: number;
  /** Number of connected groups */
  connectedGroups: number;
  /** Connection errors */
  connectionErrors: number;
}

/**
 * RabbitMQ metrics
 */
export interface RabbitMQMetrics {
  /** Number of messages in queue */
  queueDepth: number;
  /** Messages per second */
  messagesPerSecond: number;
  /** Number of consumers */
  consumers: number;
  /** Memory usage in MB */
  memoryUsageMB: number;
}

/**
 * Provider health status
 */
export interface ProviderHealthStatus {
  /** Provider type */
  providerType: ProviderType;
  /** Whether the provider is healthy */
  isHealthy: boolean;
  /** Last check time */
  lastCheckTime?: Date;
  /** Response time in milliseconds */
  responseTimeMs?: number;
  /** Error message if unhealthy */
  errorMessage?: string;
}

/**
 * Historical metrics request
 */
export interface HistoricalMetricsRequest {
  /** Start time for the query */
  startTime: Date;
  /** End time for the query */
  endTime: Date;
  /** Specific metrics to retrieve */
  metricNames: string[];
  /** Interval for data aggregation */
  interval: string;
}

/**
 * Historical metrics response
 */
export interface HistoricalMetricsResponse {
  /** Time series data */
  series: MetricSeries[];
  /** Query metadata */
  metadata: {
    startTime: Date;
    endTime: Date;
    interval: string;
    totalPoints: number;
  };
}

/**
 * Metric time series data
 */
export interface MetricSeries {
  /** Metric name */
  name: string;
  /** Data points */
  dataPoints: MetricDataPoint[];
  /** Metric metadata */
  metadata: {
    unit: string;
    description: string;
  };
}

/**
 * Individual metric data point
 */
export interface MetricDataPoint {
  /** Timestamp */
  timestamp: Date;
  /** Metric value */
  value: number;
}

/**
 * KPI summary for dashboards
 */
export interface KPISummary {
  /** Timestamp */
  timestamp: Date;
  /** System health metrics */
  systemHealth: {
    overallHealthPercentage: number;
    errorRate: number;
    responseTimeP95: number;
    activeConnections: number;
    databaseUtilization: number;
  };
  /** Performance metrics */
  performance: {
    requestsPerSecond: number;
    activeRequests: number;
    averageResponseTime: number;
    cacheHitRate: number;
  };
  /** Business metrics */
  business: {
    activeVirtualKeys: number;
    requestsPerMinute: number;
    costBurnRatePerHour: number;
    averageCostPerRequest: number;
  };
  /** Infrastructure metrics */
  infrastructure: {
    cpuUsage: number;
    memoryUsage: number;
    uptime: string;
    signalRConnections: number;
  };
}